### PR TITLE
DiscreteSystem

### PR DESCRIPTION
**Pull request by ovile27**
DiscreteSystem and DiscreteProblem functionality added.

For now, it is a simple version compared to ODESystem functionality for instance.

Test it with the example in [test/test_epidemiological_compartmental.jl](https://github.com/ovile27/ModelingToolkit.jl/blob/DiscreteSystem/test/test_epidemiological_compartmental.jl)

### CODE REVIEW

File: src/systems/discrete_system/discrete_problem.jl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    ps = parameters(sys)
+    eqs = equations(sys)
+    # defs = defaults(sys)
+    t = get_iv(sys)
+    u0 = varmap_to_vars(u0map,dvs)
+    rhss = [eq.rhs for eq in eqs]
+    u = dvs
+    p = varmap_to_vars(parammap,ps)
+    eval_module = @__MODULE__
+    eval_expression = true
```
Comment (YingboMa) [inline src/systems/discrete_system/discrete_problem.jl:17]: `eval_module` and `eval_expression` should be keyword arguments. Please check how this is done in `ODEFunction` and `ODEProblem`.

Comment (anandijain) [reply src/systems/discrete_system/discrete_problem.jl:17]: https://github.com/SciML/SciMLBase.jl/blob/master/src/problems/ode_problems.jl
https://github.com/SciML/SciMLBase.jl/blob/master/src/scimlfunctions.jl

Are these the correct links for @ovile27, Yingbo?

Comment (anandijain) [reply src/systems/discrete_system/discrete_problem.jl:17]: https://github.com/SciML/ModelingToolkit.jl/blob/8b8d549e7f5cee6467d4a032c98ce6f6f9da8623/src/systems/diffeqs/abstractodesystem.jl#L154

File: src/ModelingToolkit.jl [hunk 2]
```diff
@@ -130,6 +130,9 @@ include("systems/pde/pdesystem.jl")
 include("systems/reaction/reactionsystem.jl")
 include("systems/dependency_graphs.jl")
 
+include("systems/discrete_system/discrete_system.jl")
+include("systems/discrete_system/discrete_problem.jl")
```
Comment (ChrisRackauckas) [inline src/ModelingToolkit.jl:134]: this doesn't need two files. Just follow the structure of the other systems.

File: test/test_epidemiological_compartmental.jl [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Example: Compartmental models in epidemiology
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:1]: rename this to be the discrete system test.

File: test/test_epidemiological_compartmental.jl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+#=
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:2]: This test file is missing from runtests.jl

File: test/test_epidemiological_compartmental.jl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+- https://github.com/epirecipes/sir-julia/blob/master/markdown/function_map/function_map.md
+- https://en.wikipedia.org/wiki/Compartmental_models_in_epidemiology#Deterministic_versus_stochastic_epidemic_models
+=#
+
+using Pkg
+Pkg.activate(".") # Activate environment of ModelingToolkit.jl at DiscreteSystem branch.
+# Pkg.instantiate()
+
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:10]:
```suggestion
```

File: test/test_epidemiological_compartmental.jl [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+## We define our system based on symbolic types
+# (ModelingToolkit.jl uses Symbolics.jl internally)
+
+# Auxiliary functions
+"""
+Rate to proportion.
+
+It may be used with symbolic arguments.
+"""
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:21]:
```suggestion
```

this is just unnecessary noise.

File: test/test_epidemiological_compartmental.jl [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+using Pkg
+Pkg.activate(".") # Activate environment of ModelingToolkit.jl at DiscreteSystem branch.
+# Pkg.instantiate()
+
+using ModelingToolkit
+
+## We define our system based on symbolic types
+# (ModelingToolkit.jl uses Symbolics.jl internally)
+
+# Auxiliary functions
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:16]:
```suggestion
```

File: test/test_epidemiological_compartmental.jl [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+df_map = DataFrame(Tables.table(hcat(sol_map.u...)'))
+tmax = ModelingToolkit.value(substitute(nsteps,p)*substitute(δt,p))
+times = 0.0:ModelingToolkit.value.(substitute(δt,p)):tmax;
+df_map[!,:times] = times;
+
+@df df_map plot(:times,
+    [:Column1 :Column2 :Column3],
+    label=["S" "I" "R"],
+    xlabel="Time",
+    ylabel="Number")
```
Comment (ChrisRackauckas) [inline test/test_epidemiological_compartmental.jl:62]:
```suggestion
# Solution
using OrdinaryDiffEq
sol_map = solve(prob_map,solver=FunctionMap);
```

File: test/test_epidemiological_compartmental.jl (hunk 3 above)
Comment (ovile27) [reply test/test_epidemiological_compartmental.jl:1]: 03cb1d2663f42d09788e7440e67eb77e7b152184

File: src/ModelingToolkit.jl (hunk 2 above)
Comment (ovile27) [reply src/ModelingToolkit.jl]: 03cb1d2663f42d09788e7440e67eb77e7b152184

**Comment by YingboMa** [comment]
Remember to squash some commits before merging

**Comment by ChrisRackauckas** [comment]
I just edited that right in Github. Let's see if tests pass.

**Comment by ChrisRackauckas** [comment]
I just did a squash merge because this was messy in general.

**Pull request merged**